File: docs/guides/javascript/src/saveImageClassifications.ts
import { algoliasearch } from 'algoliasearch';

interface Image extends Record<string, unknown> {
  imageURL: string;
  objectID: string;
  objects: Record<string, unknown>[];
}

async function saveImageClassifications() {
  // Retrieve labels
  function getImageLabels(imageURL, objectID, scoreLimit): Image {
    // Implement your image classification logic here
    return { imageURL: '', objectID: '', objects: [] };
  }

  // API key ACL should include editSettings / addObject
  const client = algoliasearch('ALGOLIA_APPLICATION_ID', 'ALGOLIA_API_KEY');

  const records: Image[] = [];

  await client.browseObjects<Image>({
    indexName: '<YOUR_INDEX_NAME>',
    browseParams: undefined,
    aggregator: (browseResponse) => {
      records.push(
        ...browseResponse.hits.map((hit) => {
          return getImageLabels(hit.imageURL, hit.objectID, 0.5);
        }),
      );
    },
  });

  // Update records with image classifications
  await client.partialUpdateObjects({ indexName: 'products', objects: records, createIfNotExists: true });
}

File: docs/guides/scala/src/main/scala/saveImageClassifications.scala
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global

import algoliasearch.api.SearchClient
import algoliasearch.config.*
import algoliasearch.extension.SearchClientExtensions
import algoliasearch.search.BrowseParamsObject

class Image(
    val imageURL: String,
    val objectID: String,
    val objects: Seq[Map[String, Any]]
)

def saveImageClassifications(): Future[Unit] = {
  // Retrieve labels
  def getImageLabels(imageURL: String, objectID: String, scoreLimit: Float): Image = {
    // Implement your image classification logic here
    Image(imageURL, objectID, Seq())
  }

  // API key ACL should include editSettings / addObject
  val client = SearchClient(appId = "ALGOLIA_APPLICATION_ID", apiKey = "ALGOLIA_API_KEY")

  val records: Seq[Image] = Seq.empty

  client.browseObjects(
    indexName = "<YOUR_INDEX_NAME>",
    browseParams = BrowseParamsObject(),
    aggregator = { browseResponse =>
      records.appendedAll(
        browseResponse.hits.map { hit =>
          val props = hit.additionalProperties.getOrElse(List()).toMap
          getImageLabels(props.get("imageURL").toString, hit.objectID, 0.5f)
        }
      )
    }
  )

  // Update records with image classifications
  client
    .partialUpdateObjects(
      indexName = "<YOUR_INDEX_NAME>",
      objects = records,
      createIfNotExists = true
    )
    .map { response =>
      println(response)
    }
    .recover { case ex: Exception =>
      println(s"An error occurred: ${ex.getMessage}")
    }
}
